Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py

```diff
@@ -0,0 +1,440 @@
+    sender_value = get_str_representation_top_n_values(senders, senders_counter, top_n)
+    recipients_value = get_str_representation_top_n_values(recipients, recipients_counter, len(recipients_counter))
+    headers.append(domain_header)
+    contents.append(domain_value)
+    headers.append(sender_header)
+    contents.append(sender_value)
+    headers.append('Recipients')
+    contents.append(recipients_value)
+    for field in fields_to_display:
+        if field in incidents_df.columns:
+            field_values = get_non_na_empty_values(incidents_df, field)
+            if len(field_values) > 0:
+                field_values_counter = Counter(field_values).most_common()  # type: ignore
+                field_value_str = get_str_representation_top_n_values(field_values, field_values_counter, top_n)
+                headers.append(field)
+                contents.append(field_value_str)
+    hr = tableToMarkdown('Possible Campaign Detected', {header: value for header, value in zip(headers, contents)},
+                         headers=headers)
+    return hr
+
+
+def get_non_na_empty_values(incidents_df, field):
+    field_values = incidents_df[field].replace('', None).dropna().tolist()
+    field_values = [x for x in field_values if len(str(x).strip()) > 0]
+    return field_values
+
+
+def cosine_sim(a, b):
+    return dot(a, b) / (norm(a) * norm(b))
+
+
+def summarize_email_body(body, subject, nb_sentences=3):
+    corpus = sent_tokenize(body)
+    cv = CountVectorizer(stop_words=list(stopwords.words('english')))
+    body_arr = cv.fit_transform(corpus).toarray()
+    subject_arr = cv.transform(sent_tokenize(subject)).toarray()
+    word_list = cv.get_feature_names()
+    count_list = body_arr.sum(axis=0) + subject_arr.sum(axis=0) * 1.5
+    duplicate_sentences = [i for i, arr in enumerate(body_arr) if
+                           any(cosine_sim(arr, arr2) > DUPLICATE_SENTENCE_THRESHOLD
+                               for arr2 in body_arr[:i])]
+
+    word_frequency = dict(zip(word_list, count_list))
+    val = sorted(word_frequency.values())
+
+    max_frequency = val[-1]
+    for word in word_frequency.keys():
+        word_frequency[word] = (word_frequency[word] / max_frequency)
+    for word in KEYWORDS:
+        if word in word_frequency:
+            word_frequency[word] *= 1.5
+
+    sentence_rank = {}
+    for i, sent in enumerate(corpus):
+        if i in duplicate_sentences:
+            continue
+        sentence_rank[i] = 0
+        for word in word_tokenize(sent):
+            if word.lower() in word_frequency.keys():
+                if i in sentence_rank.keys():
+                    sentence_rank[i] += word_frequency[word.lower()]
+        sentence_rank[i] = sentence_rank[i] / len(word_tokenize(sent))  # type: ignore
+    sorted_sentence_rank = sorted(sentence_rank.items(), key=lambda item: item[1], reverse=True)
+    top_sentences_indices = [sent_i for sent_i, _ in sorted_sentence_rank[:nb_sentences]]
```

Comment:
Make sure the display order of the senteces is the same as they appeared in the original email. Here I'm not sure if the order is by the rank you gave them

Reply:
they are sorted by order later